### PR TITLE
Recent cards

### LINKED ISSUE #332: View recently obtained cards
I think it'd be nice to show recently obtained cards in either the inspector or in electron app settings under collection. 

I recently got an ICR and wanted to do some investigation on whether the game logged it and it seems that every time we obtain a card the logger prints:
`PlayerInventory.GetPlayerCardsV3(`
Afterwards it appears lists the whole collection in order of oldest acquisition to newest cards first.
(I could be wrong about this, but the last 5 cards of mine are sorted like this so I assumed it was true for everything. Probably need to test it on other ways to obtain cards besides icr's) 

Looks like this may be pretty desired on reddit too ;)
[https://www.reddit.com/r/MagicArena/search?q=recently&restrict_sr=on&sort=relevance&t=all](https://www.reddit.com/r/MagicArena/search?q=recently&restrict_sr=on&sort=relevance&t=all)

### PR DESCRIPTION
**Pull request by WizardRubic**
resolves #332
gif:
[https://i.imgur.com/bT6GEA1.gif](https://i.imgur.com/bT6GEA1.gif)

Limitations:
- Something probably needs to change in my version of electron/main.js as I went and changed the generateArgs() function so it would be okay with the spaces in my file path when building from source. (Spencer already made this change I think! This will likely require a manual merge for the file)
- The recently collected cards lists doesn't refresh unless settings are reopened so if a user collects a card while it's open they'll have to open and close settings to see it in the list.
- Most of the GRN cards don't show here either as this uses the same method as the export collection to view cards.
- For new users of MTGA tracker I wrote some code that I believe will make nothing appear here so we don't list their whole collection. It'll only list cards after MTGA tracker's first collection trigger.
- If a user collects a card while MTGA tracker isn't online, the recently collected cards list won't updated until GetPlayerCardsV3() is called in the output_log again. 
- This was implemented with a new setting tracking when and which cards are obtained. In worst case, this takes about ~4.5 times more space than storing the collection which we already do. It's probably not too consequential though unless MTGA stays around for a very very long time. It does mean the app will likely take up a bit more RAM though.  

Possible improvements in the future:
- The tracker could look through the log file on startup for GetPlayerCardsV3() calls so the user wouldn't have to collect another card to see changes on their most recent cards.
- When more card collection events should display than can fit on the screen, the scrollbar appears for the whole setting's page contents. This could be changed to have it's own mini frame so it doesn't add a big scroll bar.
- The css/design of the table could probably be improved.

### CODE REVIEW

**Comment by shawkinsl** [comment]
Sorry about Supervisor, he's just about never right. Maybe someday he'll catch a real password though :)

This looks really awesome! Re: the limitations list:

1. This is resolved in [develop/4.2.3](https://github.com/mtgatracker/mtgatracker/tree/develop/4.2.3) and will be merged soon
2. Known limitation with the entire collection. It's possible to set up some IPC to make it update without reopening, but I'm fine with leaving that for a future date. (Filed #345)
3. Also fixed in develop/4.2.3
4. Good call
5. Yeah, this is an unfortunate downside of MTGATracker as a whole right now. I think we'll be able to fix this when we finally drop python and move everything into node (see: [some warnings](https://github.com/mtgatracker/mtgatracker/blob/master/app/mtgatracker_backend.py#L145-L148))
6. I'll need to do some research on the electron settings module we're using and make sure that storing large sets of data is supported. (Or, if this is in the python side of things, we might consider reworking it to use separate rolling json files, to avoid corrupting the settings.json file)

File: electron/main.js [hunk 1]
```diff
@@ -254,6 +254,8 @@ let sortMethod = settings.get('sortMethod', 'draw');
 let useFlat = settings.get('useFlat', true);
 let useMinimal = settings.get('useMinimal', true);
 let zoom = settings.get('zoom', 0.8);
+let recentCards = settings.get('recentCards', []);
```
Comment (shawkinsl) [inline electron/main.js]: OK, good, this is using electron-settings. There isn't anything in [the docs]() warning against heavy use; only that settings shouldn't be used from the renderer process and only the main process. It looks like we pass [that check](https://github.com/mtgatracker/mtgatracker/search?q=electron-settings&unscoped_q=electron-settings)

File: electron/settingsRenderer.js [hunk 2]
```diff
@@ -164,6 +173,37 @@ rivets.binders.authref = (el, val) => {
...
 }
 
+rivets.binders.recentcardsbinder = (el, cardsObtained) => {
+  var node;
+  var textNode;
+  var currentCard
+  for(var cardID in cardsObtained) {
+    currentCard = mtga.allCards.findCard(cardID)
+    if(currentCard) {
+      textNode = document.createTextNode(cardsObtained[cardID] + "x " + currentCard.attributes.prettyName);
```
Comment (shawkinsl) [inline electron/settingsRenderer.js]:
```suggestion
      textNode = document.createTextNode(`${cardsObtained[cardID]}x ${currentCard.attributes.prettyName}`);
```

File: electron/settingsRenderer.js [hunk 3]
```diff
@@ -164,6 +173,37 @@ rivets.binders.authref = (el, val) => {
...
+    } else {
+      textNode = document.createTextNode(cardsObtained[cardID] + "x " + "card-name-not-found");
```
Comment (shawkinsl) [inline electron/settingsRenderer.js]:
```suggestion
      textNode = document.createTextNode(`${cardsObtained[cardID]}x card-name-not-found (${cardID})`);
```

Comment (shawkinsl) [reply electron/settingsRenderer.js]: (this one is more than a style change; when we don't know what the card is, at least show the grpId)

File: electron/main.js [hunk 4]
```diff
@@ -401,7 +403,7 @@ port = selectPort()
 const generateArgs = () => {
     var args = ["-p", port]
     args.push("-i")
-    args.push(logPath)
+    args.push(`"${logPath}"`)
```
[the remark this answers is not in the document]
Comment (WizardRubic) [reply electron/main.js]: I agree! 

And this GitHub feature does look sweet :p 
I'm not sure if it's cuz I pushed from cmd line, but it seems to have now marked all these convo's as outdated. Maybe I should have tried the apply suggestion feature button(I'll defo do that if I see it again)

File: electron/mainRenderer.js [hunk 5]
```diff
@@ -994,7 +998,47 @@ let onMessage = (data) => {
 //            console.log(e)
 //          })
         } else if (data.collection) {
+          // console.log("data=----------------------:");
+          // console.log(data);
+          // 
+          // 
+          // (new Date(Date.now())).toLocaleDateString()
```
[the remark this answers is not in the document]
Comment (WizardRubic) [reply electron/mainRenderer.js]: Ooops I missed a buncha of em in that file, thanks for catching it!

File: electron/settingsRenderer.js (hunk 3 above)
Comment (WizardRubic) [reply electron/settingsRenderer.js]: I forgot about the template literals, that looks much more readable than whatever I wrote haha :)

**Pull request merged**